src/firstboot/firstboot.c

```diff
@@ -317,6 +318,166 @@ static int process_locale(void) {
         return 0;
 }
 
+static int write_locale_gen(const char *localegen_path, const char *locale) {
+        _cleanup_fclose_ FILE *original = NULL, *localegen = NULL;
+        _cleanup_(unlink_and_freep) char *localegen_tmp = NULL;
+        bool found = false;
+        int r;
+
+        r = fopen_temporary_label("/etc/locale.gen", localegen_path, &localegen, &localegen_tmp);
+        if (r < 0)
+                return r;
+
+        original = fopen(localegen_path, "re");
+        if (original)
+                for (;;) {
+                        _cleanup_free_ char *line = NULL;
+                        const char *l;
+
+                        r = read_line(original, LONG_LINE_MAX, &line);
+                        if (r < 0)
+                                return r;
+                        if (r == 0)
+                                break;
+
+                        l = skip_leading_chars(line, "#");
+
+                        /* If this line isn't the locale we're looking for, write back the line
+                        unchanged. If it is the locale we're looking for, write back the line
+                        without leading whitespace and comment characters. */
+                        if (strncaseeq(locale, l, strlen(locale)))
+                                found = true;
+                        else
+                                l = line;
```

Comment:
I tried using `extract_first_word` but that broke down with the explanation comments at the start of the file. I'd get en_US.UTF-8 generated twice because it's used as an example in the beginning.

Maybe it's better to leave this logic out and overwrite the entire file instead of trying to work with the original contents? I'll drop the message about being unable to determine the charset to info instead of warning. That should give us a good enough way to determine the charset. Sure we're overwriting the entire file but if people are using the --locale option of systemd-firstboot I don't expect them to be looking at manually modifying the locale again later so maybe we don't gain all that much by trying to work with the original contents of /etc/locale.gen.

Reply:
(I dropped the logic for now. Let me know if we should re-add it)